pagination/tests/test_querystrings.py

```diff
@@ -0,0 +1,38 @@
+from django.http import QueryDict
+
+from .utils import RequestTestCase
+from .. import querystrings
+
+
+class TestQuerystringUpdate(RequestTestCase):
+    def setUp(self):
+        """Create a request and put some GET args in it."""
+        self.request = self.create_request()
+        self.request.GET = QueryDict('page=3&q=search_term')
+
+    def test_unchanged_querystring(self):
+        """Test that the method returns an unchanged request.GET if kwargs is empty."""
+        querystring = querystrings.querystring_update(self.request)
+
+        # Re-parse the result into a dict because the query arguments don't appear in
+        # a predictable order.
+        new_querydict = QueryDict(querystring).dict()
+        expected = {
+            'page': '3',
+            'q': 'search_term',
+        }
+        self.assertDictEqual(new_querydict, expected)
+
+    def test_querystring_update(self):
+        """Test that individual GET arguments are updated or preserved correctly."""
+        search_term = 'a_different_search_term'
+        querystring = querystrings.querystring_update(self.request, q=search_term)
+
+        # Re-parse the result into a dict because the query arguments don't appear in
+        # a predictable order.
+        new_querydict = QueryDict(querystring).dict()
+        expected = {
+            'page': '3',
+            'q': search_term,
+        }
+        self.assertDictEqual(new_querydict, expected)
```

Comment:
That's a good point, although in practice `assertEqual` breaks a lot so I've actually been trying to avoid it.  It fails on lists (sometimes), dictionaries (sometimes), objects (sometimes), generators (always) and basically anything more complex than a primitive (sometimes).  Worse, it leads to heisenbugs (heisentests?) that don't always fail.  TBH I'd personally prefer if we just never used it for any complex object.